test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,199 @@
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	const (
+		deviceName = "exampledevice"
+		cdiVersion = "0.2.0"
+	)
+	kind := ex.config.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+	// create CDI Files
+
+	// create spec
+	klog.Infof("Creating CDI Files")
+
+	// resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+	// validate if resourceHandle
+	if err := ex.validateResourceHandle(decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to validate resourceHandle: %v", err)
+	}
+
+	containerEditsInDevices := (&cdiapi.ContainerEdits{}).Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}})
+	for key, val := range decodedResourceHandle {
+		containerEditsInDevices.Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{
+			Env: []string{key + "=" + val},
+		}})
+	}
+
+	containerEdits := (&cdiapi.ContainerEdits{}).Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}})
+
+	spec, err := cdiapi.NewSpec(
+		&specs.Spec{
+			Version: cdiVersion,
+			Kind:    kind,
+			Devices: []specs.Device{{
+				Name:           deviceName,
+				ContainerEdits: *containerEditsInDevices.ContainerEdits,
+			}},
+			ContainerEdits: *containerEdits.ContainerEdits,
+		},
+		ex.getJsonFilePath(req.ClaimUid), 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new spec: %v", err)
+	}
+	klog.V(5).InfoS("CDI spec is validated")
+
+	// create json file from spec
+	doc, err := json.Marshal(*spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	if err := os.WriteFile(ex.getJsonFilePath(req.ClaimUid), doc, os.FileMode(0644)); err != nil {
+		return nil, fmt.Errorf("failed to write file: %v", err)
+	}
+	klog.Infof("Json file is created at " + ex.getJsonFilePath(req.ClaimUid))
+
+	//return qualified name of the cdi device
+	dev := spec.GetDevice(deviceName).GetQualifiedName()
+	klog.Infof("Return: " + dev)
+	return &drapbv1.NodePrepareResourceResponse{CdiDevice: []string{dev}}, nil
+}
+
+func (ex *examplePlugin) NodeUnprepareResource(ctx context.Context, req *drapbv1.NodeUnprepareResourceRequest) (*drapbv1.NodeUnprepareResourceResponse, error) {
+	klog.Infof("NodeUnprepareResource is called: request: %+v", req)
+	if _, err := os.Stat(ex.getJsonFilePath(req.ClaimUid)); err == nil {
+		// jsonFile exists
+		os.Remove(ex.getJsonFilePath(req.ClaimUid))
+		klog.Infof("Json file " + ex.getJsonFilePath(req.ClaimUid) + " is removed")
+	} else if errors.Is(err, os.ErrNotExist) {
+		// jsonFile does not exist
+		klog.Infof("Json file " + ex.getJsonFilePath(req.ClaimUid) + " is already deleted")
```

Comment:
That should be a warning I believe. The fact that it's deleted signals that something went wrong.

Reply:
I'd suggest not to overcomplicate things here. It's an example code. I'd do something like this:
```
if err := WriteFile(path); err != nil {
    return nil, fmt.Errorf("error creating CDI file %s: %v", path, err)
}
```
err will contain more info about the reason.

Reply:
Not resolved either?

Reply:
But if it is marked resolved without me checking it first, then I won't know that I still need to check it because GitHub hides the discussion thread.